runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java

```diff
@@ -0,0 +1,254 @@
+            .setPipeline(PipelineTranslation.toProto(pipeline))
+            .setPipelineOptions(PipelineOptionsTranslation.toProto(options))
+            .build();
+
+    ManagedChannel jobServiceChannel =
+        channelFactory.forDescriptor(getApiServiceDescriptor(endpoint));
+
+    JobServiceBlockingStub jobService = JobServiceGrpc.newBlockingStub(jobServiceChannel);
+    CloseableResource<JobServiceBlockingStub> wrappedJobService =
+        CloseableResource.of(jobService, (unused) -> jobServiceChannel.shutdown());
+
+    PrepareJobResponse prepareJobResponse = jobService.prepare(prepareJobRequest);
+    LOG.info("PrepareJobResponse: {}", prepareJobResponse);
+
+    ApiServiceDescriptor artifactStagingEndpoint = prepareJobResponse.getArtifactStagingEndpoint();
+
+    String stagingToken = null;
+    try (CloseableResource<ManagedChannel> artifactChannel =
+        CloseableResource.of(
+            channelFactory.forDescriptor(artifactStagingEndpoint), ManagedChannel::shutdown)) {
+      ArtifactServiceStager stager = ArtifactServiceStager.overChannel(artifactChannel.get());
+      LOG.debug("Actual files staged: {}", filesToStage);
+      stagingToken = stager.stage(filesToStage);
+    } catch (CloseableResource.CloseException e) {
+      LOG.warn("Error closing artifact staging channel", e);
+      // CloseExceptions should only be thrown while closing the channel.
+      checkState(stagingToken != null);
+    } catch (Exception e) {
+      throw new RuntimeException("Error staging files.", e);
+    }
+
+    RunJobRequest runJobRequest =
+        RunJobRequest.newBuilder()
+            .setPreparationId(prepareJobResponse.getPreparationId())
+            .setStagingToken(stagingToken)
+            .build();
+
+    RunJobResponse runJobResponse = jobService.run(runJobRequest);
+
+    LOG.info("RunJobResponse: {}", runJobResponse);
+    ByteString jobId = runJobResponse.getJobIdBytes();
+
+    return new JobServicePipelineResult(jobId, wrappedJobService);
+  }
+
+  private static Endpoints.ApiServiceDescriptor getApiServiceDescriptor(String descriptor) {
+    Endpoints.ApiServiceDescriptor.Builder apiServiceDescriptorBuilder =
+        Endpoints.ApiServiceDescriptor.newBuilder();
+    apiServiceDescriptorBuilder.setUrl(descriptor);
+    return apiServiceDescriptorBuilder.build();
+  }
+
+  private void replaceTransforms(Pipeline pipeline, boolean streaming) {
+    pipeline.replaceAll(getOverrides(streaming));
+  }
+
+  private List<PTransformOverride> getOverrides(boolean streaming) {
+    return ImmutableList.of(JavaReadViaImpulse.boundedOverride());
+  }
+
+  @Override
+  public String toString() {
+    return "PortableRunner#" + hashCode();
+  }
+
+  private static ManagedChannelFactory getChannelFactory(PipelineOptions options) {
+    ManagedChannelFactory channelFactory;
+    List<String> experiments = options.as(ExperimentalOptions.class).getExperiments();
+    if (experiments != null && experiments.contains("beam_fn_api_epoll")) {
```

Comment:
<!--thread_id:cc_182207885_t; commit:d5c571bb36cf1f18d098b82592729e85d9037059; resolved:0-->
<!--section:context-quote-->
> **bsidhom** wrote:
> Not sure. I assume this was copied from https://github.com/apache/beam/blob/15da6412fef89dc08de59828a5df9ba11e944a79/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java#L100
> 
> If we don't have a reason to keep this, we can always use the default factory.

<!--section:body-->
I'd suggest to remove this. This seems like a really old and forgotten experiment, perhaps something Luke was using once for performance testing, and I don't think anybody ever turns it on.